File: packages/shared/src/energy.ts
import type { EnergyInputs, EnergyOutputs } from "./types";

const round = (value: number) => Number(value.toFixed(2));
const ratio = (numerator: number, denominator: number) =>
  denominator > 0 ? numerator / denominator : 0;

export function calculateEnergy(input: EnergyInputs): EnergyOutputs {
  const dailyGeneration = input.arrayKw * input.peakSunHours * input.systemEfficiency;
  const E_gen = dailyGeneration * 30;

  const dailyDemand = input.monthlyDemandKwh / 30;
  const dailyDaytimeDemand = dailyDemand * input.daytimeDemandFraction;
  const dailyNightDemand = dailyDemand - dailyDaytimeDemand;

  const directDaily = Math.min(dailyGeneration, dailyDaytimeDemand);
  const excessDaily = Math.max(0, dailyGeneration - directDaily);
  const usableBattery = input.batteryKwh * input.batteryDepthOfDischarge;
  const chargedDaily = Math.min(excessDaily, usableBattery);
  const batteryUsedDaily = Math.min(
    chargedDaily * input.batteryRoundTripEfficiency,
    dailyNightDemand,
  );
  const wasteDaily = Math.max(0, excessDaily - chargedDaily);

  const E_direct = directDaily * 30;
  const E_charge = chargedDaily * 30;
  const E_battery_used = batteryUsedDaily * 30;
  const E_gen_raw = round(E_gen);
  const E_sold_uncapped = round(E_direct + E_battery_used);
  const E_sold = round(Math.min(E_sold_uncapped, input.monthlyDemandKwh, E_gen_raw));
  const E_waste = round(Math.max(0, E_gen_raw - E_direct - E_charge));
  const E_grid = round(Math.max(0, input.monthlyDemandKwh - E_sold));

  const utilization = round(ratio(E_sold, E_gen_raw));
  const wasteRate = round(ratio(E_waste, E_gen_raw));

  return {
    E_gen: E_gen_raw,
    E_direct: round(E_direct),
    E_charge: round(E_charge),
    E_battery_used: round(E_battery_used),
    E_sold,
    E_waste,
    E_grid,
    utilization,
    wasteRate,
    coverage: round(ratio(E_sold, input.monthlyDemandKwh)),
  };
}

export function calculateRevenue(E_sold: number, pricePerKwh: number) {
  return round(E_sold * pricePerKwh);
}

export function calculateSavings(E_sold: number, gridPriceKes: number, solarPriceKes: number) {
  return round(Math.max(0, (gridPriceKes - solarPriceKes) * E_sold));
}

File: backend/app/services/energy.py
def round2(value: float) -> float:
    return round(value + 1e-12, 2)


def ratio(numerator: float, denominator: float) -> float:
    return numerator / denominator if denominator > 0 else 0


def calculate_energy(input: dict) -> dict:
    daily_generation = input["arrayKw"] * input["peakSunHours"] * input["systemEfficiency"]
    daily_demand = input["monthlyDemandKwh"] / 30
    daily_daytime_demand = daily_demand * input["daytimeDemandFraction"]
    daily_night_demand = daily_demand - daily_daytime_demand
    direct_daily = min(daily_generation, daily_daytime_demand)
    excess_daily = max(0, daily_generation - direct_daily)
    usable_battery = input["batteryKwh"] * input["batteryDepthOfDischarge"]
    charged_daily = min(excess_daily, usable_battery)
    battery_used_daily = min(charged_daily * input["batteryRoundTripEfficiency"], daily_night_demand)
    waste_daily = max(0, excess_daily - charged_daily)

    e_gen = round2(daily_generation * 30)
    e_direct = round2(direct_daily * 30)
    e_charge = round2(charged_daily * 30)
    e_battery_used = round2(battery_used_daily * 30)
    e_sold_uncapped = round2(e_direct + e_battery_used)
    e_sold = round2(min(e_sold_uncapped, input["monthlyDemandKwh"], e_gen))
    e_waste = round2(max(0, e_gen - e_direct - e_charge))
    e_grid = round2(max(0, input["monthlyDemandKwh"] - e_sold))
    utilization = round2(ratio(e_sold, e_gen))
    waste_rate = round2(ratio(e_waste, e_gen))
    return {
        "E_gen": e_gen,
        "E_direct": e_direct,
        "E_charge": e_charge,
        "E_battery_used": e_battery_used,
        "E_sold": e_sold,
        "E_waste": e_waste,
        "E_grid": e_grid,
        "utilization": utilization,
        "wasteRate": waste_rate,
        "coverage": round2(ratio(e_sold, input["monthlyDemandKwh"])),
    }


def calculate_revenue(e_sold: float, price_per_kwh: float) -> float:
    return round2(e_sold * price_per_kwh)


def calculate_savings(e_sold: float, grid_price_kes: float, solar_price_kes: float) -> float:
    return round2(max(0, (grid_price_kes - solar_price_kes) * e_sold))
